File: backend/app/core/partner_offer_readiness.py
"""Partner offer territorial readiness — RF-02A (no YM spend)."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from enum import StrEnum
from typing import Any, Literal

from app.core.organization_constants import OrganizationVisibility
from app.core.partner_constants import PUBLIC_PARTNER_STATUSES, PartnerStatus
from app.core.partner_offer_value_categories import (
    PartnerOfferValueCategory,
    infer_partner_offer_value_category,
    partner_offer_value_category_label,
)
from app.core.passport_constants import PartnerOfferStatus, PartnerOfferType

PLACEHOLDER_DESCRIPTION_SNIPPET = "présentez votre passport yunicity pour découvrir"
PLACEHOLDER_CONDITIONS_SNIPPET = "offre pilote, modalités confirmées sur place"

VAGUE_VALUE_LABELS: frozenset[str] = frozenset(
    {
        "avantage membre",
        "offre pilote",
        "découverte",
        "offre partenaire",
        "avantage passport",
    }
)


class PartnerOfferReadinessStatus(StrEnum):
    READY = "ready"
    PARTIAL = "partial"
    NOT_READY = "not_ready"


CheckSeverity = Literal["ok", "warning", "error"]


@dataclass(frozen=True)
class PartnerOfferReadinessCheck:
    key: str
    label: str
    passed: bool
    severity: CheckSeverity


@dataclass(frozen=True)
class PartnerOfferReadinessResult:
    status: PartnerOfferReadinessStatus
    is_passport_eligible: bool
    is_placeholder: bool
    value_category: PartnerOfferValueCategory
    value_category_label: str
    checks: tuple[PartnerOfferReadinessCheck, ...]


@dataclass(frozen=True)
class PartnerOfferReadinessInput:
    title: str
    description: str | None
    value_label: str | None
    conditions: str | None
    offer_type: PartnerOfferType | str
    offer_status: PartnerOfferStatus | str
    is_active: bool
    valid_from: datetime | None
    valid_until: datetime | None
    partner_status: str | None
    org_visibility: str | None
    org_verified: bool
    metadata: dict[str, Any] | None = None


def _normalize_status(value: PartnerOfferStatus | str) -> PartnerOfferStatus:
    if isinstance(value, PartnerOfferStatus):
        return value
    return PartnerOfferStatus(value)


def _text(value: str | None) -> str:
    return (value or "").strip()


def is_partner_offer_placeholder(
    *,
    title: str,
    description: str | None,
    value_label: str | None,
    conditions: str | None,
) -> bool:
    description_lower = _text(description).lower()
    conditions_lower = _text(conditions).lower()
    value_lower = _text(value_label).lower()
    title_lower = _text(title).lower()

    if PLACEHOLDER_DESCRIPTION_SNIPPET in description_lower:
        return True
    if PLACEHOLDER_CONDITIONS_SNIPPET in conditions_lower:
        return True
    if value_lower in VAGUE_VALUE_LABELS and (
        PLACEHOLDER_DESCRIPTION_SNIPPET in description_lower
        or len(_text(description)) < 25
    ):
        return True
    if title_lower in {"accueil passport", "avantage passport", "avantage membre yunicity"}:
        return True
    return False


def _dates_valid(
    *,
    valid_from: datetime | None,
    valid_until: datetime | None,
    now: datetime,
) -> bool:
    if valid_from is not None and valid_from > now:
        return False
    if valid_until is not None and valid_until < now:
        return False
    return True


def _partner_is_active(partner_status: str | None) -> bool:
    if not partner_status:
        return False
    try:
        status = PartnerStatus(partner_status)
    except ValueError:
        return False
    return status in PUBLIC_PARTNER_STATUSES


def _benefit_defined(
    *,
    description: str | None,
    value_label: str | None,
    is_placeholder: bool,
) -> bool:
    if is_placeholder:
        return False
    value = _text(value_label)
    if len(value) >= 4 and value.lower() not in VAGUE_VALUE_LABELS:
        return True
    description_text = _text(description)
    return len(description_text) >= 20 and PLACEHOLDER_DESCRIPTION_SNIPPET not in (
        description_text.lower()
    )


def _conditions_defined(*, conditions: str | None, is_placeholder: bool) -> bool:
    if is_placeholder:
        return False
    text = _text(conditions)
    if len(text) < 10:
        return False
    return PLACEHOLDER_CONDITIONS_SNIPPET not in text.lower()


def partner_offer_human_description(
    *,
    title: str,
    value_label: str | None,
    description: str | None,
    conditions: str | None,
    value_category: PartnerOfferValueCategory | None = None,
    offer_type: PartnerOfferType | str | None = None,
    metadata: dict[str, Any] | None = None,
) -> str:
    category = value_category
    if category is None and offer_type is not None:
        category = infer_partner_offer_value_category(
            offer_type=offer_type,
            value_label=value_label,
            title=title,
            metadata=metadata,
        )
    benefit = _text(value_label) or _text(description) or _text(title)
    category_label = partner_offer_value_category_label(category) if category else "Avantage"
    base = f"{category_label} — {benefit}"
    cond = _text(conditions)
    if cond:
        return f"{base}. {cond}"
    return base


def partner_offer_readiness(
    data: PartnerOfferReadinessInput,
    *,
    now: datetime | None = None,
) -> PartnerOfferReadinessResult:
    moment = now or datetime.now(UTC)
    status = _normalize_status(data.offer_status)
    is_placeholder = is_partner_offer_placeholder(
        title=data.title,
        description=data.description,
        value_label=data.value_label,
        conditions=data.conditions,
    )
    value_category = infer_partner_offer_value_category(
        offer_type=data.offer_type,
        value_label=data.value_label,
        title=data.title,
        metadata=data.metadata,
    )

    title_ok = len(_text(data.title)) >= 3
    benefit_ok = _benefit_defined(
        description=data.description,
        value_label=data.value_label,
        is_placeholder=is_placeholder,
    )
    conditions_ok = _conditions_defined(
        conditions=data.conditions,
        is_placeholder=is_placeholder,
    )
    visibility_ok = status == PartnerOfferStatus.PUBLISHED and data.is_active
    partner_ok = _partner_is_active(data.partner_status)
    org_public = data.org_visibility == OrganizationVisibility.PUBLIC.value
    org_verified = data.org_verified
    dates_ok = _dates_valid(
        valid_from=data.valid_from,
        valid_until=data.valid_until,
        now=moment,
    )

    checks = (
        PartnerOfferReadinessCheck(
            key="title_defined",
            label="Titre défini",
            passed=title_ok,
            severity="ok" if title_ok else "error",
        ),
        PartnerOfferReadinessCheck(
            key="benefit_defined",
            label="Avantage défini",
            passed=benefit_ok,
            severity="ok" if benefit_ok else "error",
        ),
        PartnerOfferReadinessCheck(
            key="conditions_defined",
            label="Conditions définies",
            passed=conditions_ok,
            severity="ok" if conditions_ok else "warning",
        ),
        PartnerOfferReadinessCheck(
            key="visibility_enabled",
            label="Visibilité activée",
            passed=visibility_ok,
            severity="ok" if visibility_ok else "warning",
        ),
        PartnerOfferReadinessCheck(
            key="partner_active",
            label="Partenaire actif",
            passed=partner_ok and org_verified,
            severity="ok" if partner_ok and org_verified else "error",
        ),
        PartnerOfferReadinessCheck(
            key="dates_valid",
            label="Dates de validité",
            passed=dates_ok,
            severity="ok" if dates_ok else "warning",
        ),
        PartnerOfferReadinessCheck(
            key="not_placeholder",
            label="Contenu réel (non placeholder)",
            passed=not is_placeholder,
            severity="ok" if not is_placeholder else "error",
        ),
    )

    core_ready = (
        title_ok
        and benefit_ok
        and conditions_ok
        and not is_placeholder
        and partner_ok
        and org_verified
    )

    if core_ready and visibility_ok and dates_ok and org_public:
        readiness_status = PartnerOfferReadinessStatus.READY
    elif is_placeholder or not title_ok or not partner_ok:
        readiness_status = PartnerOfferReadinessStatus.NOT_READY
    elif core_ready or (title_ok and benefit_ok):
        readiness_status = PartnerOfferReadinessStatus.PARTIAL
    else:
        readiness_status = PartnerOfferReadinessStatus.NOT_READY

    is_passport_eligible = (
        readiness_status == PartnerOfferReadinessStatus.READY
        and visibility_ok
        and dates_ok
        and org_public
        and not is_placeholder
    )

    return PartnerOfferReadinessResult(
        status=readiness_status,
        is_passport_eligible=is_passport_eligible,
        is_placeholder=is_placeholder,
        value_category=value_category,
        value_category_label=partner_offer_value_category_label(value_category),
        checks=checks,
    )

File: frontend/packages/utils/src/partner-offer-readiness.ts
/** Partner offer value categories — RF-02A */

import type { PartnerOfferType } from "@yunicity/types";

export type PartnerOfferValueCategory =
  | "percent_discount"
  | "free_item"
  | "loyalty_reward"
  | "exclusive_access"
  | "event_benefit"
  | "welcome_bonus";

export type PartnerOfferReadinessLevel = "ready" | "partial" | "not_ready";

export type PartnerOfferReadinessCheckSeverity = "ok" | "warning" | "error";

export interface PartnerOfferReadinessCheck {
  key: string;
  label: string;
  passed: boolean;
  severity: PartnerOfferReadinessCheckSeverity;
}

export interface PartnerOfferReadinessFields {
  readiness: PartnerOfferReadinessLevel;
  is_passport_eligible: boolean;
  is_placeholder: boolean;
  value_category: PartnerOfferValueCategory;
  value_category_label: string;
  human_description: string;
  checks: PartnerOfferReadinessCheck[];
}

export const PARTNER_OFFER_VALUE_CATEGORY_LABELS: Record<PartnerOfferValueCategory, string> = {
  percent_discount: "Réduction",
  free_item: "Produit offert",
  loyalty_reward: "Récompense fidélité",
  exclusive_access: "Accès exclusif",
  event_benefit: "Avantage événement",
  welcome_bonus: "Bonus bienvenue",
};

export const PARTNER_OFFER_READINESS_LABELS: Record<PartnerOfferReadinessLevel, string> = {
  ready: "Prête",
  partial: "Partielle",
  not_ready: "Non prête",
};

const PLACEHOLDER_DESCRIPTION_SNIPPET = "présentez votre passport yunicity pour découvrir";
const PLACEHOLDER_CONDITIONS_SNIPPET = "offre pilote, modalités confirmées sur place";

const VAGUE_VALUE_LABELS = new Set([
  "avantage membre",
  "offre pilote",
  "découverte",
  "offre partenaire",
  "avantage passport",
]);

const DEFAULT_CATEGORY_BY_OFFER_TYPE: Record<PartnerOfferType, PartnerOfferValueCategory> = {
  drink: "free_item",
  discount: "percent_discount",
  vip: "exclusive_access",
  gift: "free_item",
  event_access: "event_benefit",
  custom: "exclusive_access",
};

export function partnerOfferValueCategoryLabel(category: PartnerOfferValueCategory): string {
  return PARTNER_OFFER_VALUE_CATEGORY_LABELS[category] ?? "Avantage";
}

export function partnerOfferReadinessLabel(level: PartnerOfferReadinessLevel): string {
  return PARTNER_OFFER_READINESS_LABELS[level];
}

export function inferPartnerOfferValueCategory(input: {
  offer_type: PartnerOfferType;
  value_label?: string | null;
  title?: string | null;
  metadata?: Record<string, unknown> | null;
}): PartnerOfferValueCategory {
  const raw = input.metadata?.value_category;
  if (typeof raw === "string" && raw in PARTNER_OFFER_VALUE_CATEGORY_LABELS) {
    return raw as PartnerOfferValueCategory;
  }

  const haystack = [input.value_label, input.title]
    .filter((part): part is string => Boolean(part?.trim()))
    .join(" ")
    .toLowerCase();

  if (haystack.includes("bienvenue") || haystack.includes("première visite")) {
    return "welcome_bonus";
  }
  if (haystack.includes("fidélité") || haystack.includes("fidelite")) {
    return "loyalty_reward";
  }
  if (haystack.includes("%") || haystack.includes("réduction") || haystack.includes("reduction")) {
    return "percent_discount";
  }

  return DEFAULT_CATEGORY_BY_OFFER_TYPE[input.offer_type] ?? "exclusive_access";
}

export function isPartnerOfferPlaceholder(input: {
  title: string;
  description?: string | null;
  value_label?: string | null;
  conditions?: string | null;
}): boolean {
  const description = (input.description ?? "").trim().toLowerCase();
  const conditions = (input.conditions ?? "").trim().toLowerCase();
  const valueLabel = (input.value_label ?? "").trim().toLowerCase();
  const title = (input.title ?? "").trim().toLowerCase();

  if (description.includes(PLACEHOLDER_DESCRIPTION_SNIPPET)) return true;
  if (conditions.includes(PLACEHOLDER_CONDITIONS_SNIPPET)) return true;
  if (
    VAGUE_VALUE_LABELS.has(valueLabel) &&
    (description.includes(PLACEHOLDER_DESCRIPTION_SNIPPET) || description.length < 25)
  ) {
    return true;
  }
  if (["accueil passport", "avantage passport", "avantage membre yunicity"].includes(title)) {
    return true;
  }
  return false;
}

export function partnerOfferHumanDescription(input: {
  title: string;
  value_label?: string | null;
  description?: string | null;
  conditions?: string | null;
  offer_type?: PartnerOfferType;
  value_category?: PartnerOfferValueCategory;
  metadata?: Record<string, unknown> | null;
}): string {
  const category =
    input.value_category ??
    (input.offer_type
      ? inferPartnerOfferValueCategory({
          offer_type: input.offer_type,
          value_label: input.value_label,
          title: input.title,
          metadata: input.metadata,
        })
      : "exclusive_access");
  const benefit =
    input.value_label?.trim() || input.description?.trim() || input.title.trim();
  const base = `${partnerOfferValueCategoryLabel(category)} — ${benefit}`;
  const conditions = input.conditions?.trim();
  return conditions ? `${base}. ${conditions}` : base;
}

export interface PartnerOfferReadinessInput {
  title: string;
  description?: string | null;
  value_label?: string | null;
  conditions?: string | null;
  offer_type: PartnerOfferType;
  offer_status: string;
  is_active: boolean;
  valid_from?: string | null;
  valid_until?: string | null;
  partner_status?: string | null;
  org_visibility?: string | null;
  org_verified?: boolean;
  metadata?: Record<string, unknown> | null;
}

function datesValid(
  validFrom: string | null | undefined,
  validUntil: string | null | undefined,
  now: Date,
): boolean {
  if (validFrom) {
    const start = new Date(validFrom);
    if (!Number.isNaN(start.getTime()) && start.getTime() > now.getTime()) return false;
  }
  if (validUntil) {
    const end = new Date(validUntil);
    if (!Number.isNaN(end.getTime()) && end.getTime() < now.getTime()) return false;
  }
  return true;
}

function partnerIsActive(partnerStatus: string | null | undefined): boolean {
  if (!partnerStatus) return false;
  return ["active", "premium", "founding_partner"].includes(partnerStatus);
}

export function partnerOfferReadiness(
  input: PartnerOfferReadinessInput,
  now = new Date(),
): PartnerOfferReadinessFields {
  const placeholder = isPartnerOfferPlaceholder(input);
  const valueCategory = inferPartnerOfferValueCategory({
    offer_type: input.offer_type,
    value_label: input.value_label,
    title: input.title,
    metadata: input.metadata,
  });

  const titleOk = input.title.trim().length >= 3;
  const value = (input.value_label ?? "").trim();
  const description = (input.description ?? "").trim();
  const benefitOk =
    !placeholder &&
    ((value.length >= 4 && !VAGUE_VALUE_LABELS.has(value.toLowerCase())) ||
      (description.length >= 20 && !description.toLowerCase().includes(PLACEHOLDER_DESCRIPTION_SNIPPET)));
  const conditionsText = (input.conditions ?? "").trim();
  const conditionsOk =
    !placeholder &&
    conditionsText.length >= 10 &&
    !conditionsText.toLowerCase().includes(PLACEHOLDER_CONDITIONS_SNIPPET);
  const visibilityOk = input.offer_status === "published" && input.is_active;
  const partnerOk = partnerIsActive(input.partner_status) && Boolean(input.org_verified);
  const orgPublic = input.org_visibility === "public";
  const datesOk = datesValid(input.valid_from, input.valid_until, now);

  const checks: PartnerOfferReadinessCheck[] = [
    {
      key: "title_defined",
      label: "Titre défini",
      passed: titleOk,
      severity: titleOk ? "ok" : "error",
    },
    {
      key: "benefit_defined",
      label: "Avantage défini",
      passed: benefitOk,
      severity: benefitOk ? "ok" : "error",
    },
    {
      key: "conditions_defined",
      label: "Conditions définies",
      passed: conditionsOk,
      severity: conditionsOk ? "ok" : "warning",
    },
    {
      key: "visibility_enabled",
      label: "Visibilité activée",
      passed: visibilityOk,
      severity: visibilityOk ? "ok" : "warning",
    },
    {
      key: "partner_active",
      label: "Partenaire actif",
      passed: partnerOk,
      severity: partnerOk ? "ok" : "error",
    },
    {
      key: "dates_valid",
      label: "Dates de validité",
      passed: datesOk,
      severity: datesOk ? "ok" : "warning",
    },
    {
      key: "not_placeholder",
      label: "Contenu réel (non placeholder)",
      passed: !placeholder,
      severity: placeholder ? "error" : "ok",
    },
  ];

  const coreReady = titleOk && benefitOk && conditionsOk && !placeholder && partnerOk;
  let readiness: PartnerOfferReadinessLevel;
  if (coreReady && visibilityOk && datesOk && orgPublic) {
    readiness = "ready";
  } else if (placeholder || !titleOk || !partnerOk) {
    readiness = "not_ready";
  } else if (coreReady || (titleOk && benefitOk)) {
    readiness = "partial";
  } else {
    readiness = "not_ready";
  }

  const isPassportEligible =
    readiness === "ready" && visibilityOk && datesOk && orgPublic && !placeholder;

  return {
    readiness,
    is_passport_eligible: isPassportEligible,
    is_placeholder: placeholder,
    value_category: valueCategory,
    value_category_label: partnerOfferValueCategoryLabel(valueCategory),
    human_description: partnerOfferHumanDescription({
      title: input.title,
      value_label: input.value_label,
      description: input.description,
      conditions: input.conditions,
      offer_type: input.offer_type,
      value_category: valueCategory,
      metadata: input.metadata,
    }),
    checks,
  };
}

/** Re-export type label helper used by public catalog. */
export { partnerOfferTypeLabel, partnerOfferValueLabel } from "./partner-offer-public";
